cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentConsumer.java

```diff
@@ -0,0 +1,108 @@
+ *            return null;
+ *          }
+ *
+ *          final MessageDigest messageDigest = MessageDigest.getInstance(digestType);
+ *          return new HttpContentConsumer() {
+ *
+ *            {@literal@}Override
+ *            public void onReceived(ByteBuffer chunk) throws Exception {
+ *              messageDigest.update(chunk);
+ *            }
+ *
+ *            {@literal@}Override
+ *            public void onFinish(HttpServiceResponder responder) throws Exception {
+ *              responder.sendString(Bytes.toHexString(messageDigest.digest()));
+ *            }
+ *
+ *            {@literal@}Override
+ *            public void onError(HttpServiceResponder responder, Throwable failureCause) {
+ *              responder.sendError(500, failureCause.getMessage());
+ *            }
+ *          }
+ *        }
+ *      }
+ *   </code></pre>
+ * </p>
+ */
+public abstract class HttpContentConsumer {
+
+  /**
+   * This method will get invoked when a new chunk of the request body is available to be consumed.
+   * It is guaranteed that no concurrent calls to this method will be made.
+   * <p>
+   * Also note that each invocation
+   * to this method is executed inside a single transaction, hence access to transactional {@link Dataset}
+   * is possible through the {@link HttpServiceContext} provided to the
+   * {@link HttpServiceHandler#initialize(HttpServiceContext)} method.
+   * </p>
+   *
+   * If transaction is not always needed, you can extend from the {@link NonTransactionalHttpContentConsumer} class and
+   * implement the {@link NonTransactionalHttpContentConsumer#onReceived(ByteBuffer, Transactional)} method instead
+   * to get better performance.
+   *
+   * @param chunk a {@link ByteBuffer} containing a chunk of the request body
+   * @throws Exception if there is any error when processing the received chunk
+   */
+  public abstract void onReceived(ByteBuffer chunk) throws Exception;
+
+  /**
+   * This method will get invoked when reached the end of the request body. It must use the given
+   * {@link HttpServiceResponder} to send respond in order to complete the HTTP call. This method is
+   * always executed inside a single transaction.
+   *
+   * @param responder a {@link HttpServiceResponder} for sending respond
+   * @throws Exception if there is any error
+   */
+  public abstract void onFinish(HttpServiceResponder responder) throws Exception;
+
+  /**
+   * This method will get invoked when there is error while processing the request body chunks. If {@link Exception}
+   * are being thrown from either the {@link #onReceived(ByteBuffer)} or {@link #onFinish(HttpServiceResponder)}
+   * methods, this method will get called. This method is always executed inside a single transaction.
+   *
+   * @param responder a {@link HttpServiceResponder} for sending respond
+   * @param failureCause the reason of the failure
+   */
+  public abstract void onError(HttpServiceResponder responder, Throwable failureCause);
+}
```

Comment:
It was relying on the netty-http library to call `BodyConsumer.handleError` when there is error so that the `BodyConsumerAdapter` (down below) will do the forwarding. However, it seems like there is bug on the netty-http library side that when body consumer is involve, nothing will get triggered if the client close connect before completing the streaming.

Will fix it separate from the netty-http library to guarantee right calls are being made.